File: ConstructBinaryTreeFromPreorderAndInorderTraversal/ConstructBinaryTreeFromPreorderAndInorderTraversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def buildTree(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: TreeNode
        """
        if len(preorder) == 0:
            return None
        pleft = 0
        pright = len(preorder) - 1
        ileft = 0
        iright = len(inorder) - 1
        return build2(preorder, inorder, pleft, pright, ileft, iright)

def build2(preorder, inorder, pleft, pright, ileft, iright):
    if pleft > pright or ileft > iright:
        return
    if pleft == pright:
        return TreeNode(preorder[pleft])
    root = TreeNode(preorder[pleft])
    index = inorder.index(preorder[pleft])
    index = index - ileft
    left = build2(preorder, inorder, pleft + 1, pleft + index, ileft, ileft + index - 1)
    right = build2(preorder, inorder, pleft + index + 1, pright, ileft + index + 1, iright)
    root.left = left
    root.right = right
    return root

File: ConstructBinaryTreeFromPreorderAndInorderTraversal/ConstructBinaryTreeFromPreorderAndInorderTraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        unordered_map<int, int> inorder_map;
        for(int i = 0; i < inorder.size(); i++) {
        	inorder_map[inorder[i]] = i;
        }
        TreeNode* root =  build(preorder, 0, preorder.size() - 1, 0, inorder.size() - 1, inorder_map);
        return root;
    }
private:
	TreeNode* build(vector<int>& preorder, int p_begin, int p_end, int i_begin, int i_end, unordered_map<int, int>& inorder_map) {
    	if(i_begin > i_end || p_begin > p_end)
    		return nullptr;
    	TreeNode* node = new TreeNode(preorder[p_begin]);
    	int index = inorder_map.find(preorder[p_begin])->second;
    	int l = index - i_begin;
    	node->left = build(preorder, p_begin + 1, p_begin + l, i_begin, i_begin + l - 1, inorder_map);
    	node->right = build(preorder, p_begin + l + 1, p_end, i_begin + l + 1, i_end, inorder_map);
    	return node;
    }
};